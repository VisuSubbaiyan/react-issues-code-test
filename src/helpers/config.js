export const theme = {
  primary: '#20232a',
  secondary: '#ccc',
  black: '#282c34',
  lightGrey: '#dcdcdc',
  smallFont: '10px',
  linkColor: '#0366d6',
  breakpoints: {
    small: 640,
    medium: 769
  }
};
