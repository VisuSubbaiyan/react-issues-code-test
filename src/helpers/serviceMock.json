[
  {
    "url": "https://api.github.com/repos/facebook/react/issues/14974",
    "repository_url": "https://api.github.com/repos/facebook/react",
    "labels_url": "https://api.github.com/repos/facebook/react/issues/14974/labels{/name}",
    "comments_url": "https://api.github.com/repos/facebook/react/issues/14974/comments",
    "events_url": "https://api.github.com/repos/facebook/react/issues/14974/events",
    "html_url": "https://github.com/facebook/react/pull/14974",
    "id": 415565783,
    "node_id": "MDExOlB1bGxSZXF1ZXN0MjU3MDIzNTU3",
    "number": 14974,
    "title": "Fix act() tests by properly unmounting the container",
    "user": {
      "login": "philipp-spiess",
      "id": 458591,
      "node_id": "MDQ6VXNlcjQ1ODU5MQ==",
      "avatar_url": "https://avatars0.githubusercontent.com/u/458591?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/philipp-spiess",
      "html_url": "https://github.com/philipp-spiess",
      "followers_url": "https://api.github.com/users/philipp-spiess/followers",
      "following_url": "https://api.github.com/users/philipp-spiess/following{/other_user}",
      "gists_url": "https://api.github.com/users/philipp-spiess/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/philipp-spiess/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/philipp-spiess/subscriptions",
      "organizations_url": "https://api.github.com/users/philipp-spiess/orgs",
      "repos_url": "https://api.github.com/users/philipp-spiess/repos",
      "events_url": "https://api.github.com/users/philipp-spiess/events{/privacy}",
      "received_events_url": "https://api.github.com/users/philipp-spiess/received_events",
      "type": "User",
      "site_admin": false
    },
    "labels": [
      {
        "id": 196858374,
        "node_id": "MDU6TGFiZWwxOTY4NTgzNzQ=",
        "url": "https://api.github.com/repos/facebook/react/labels/CLA%20Signed",
        "name": "CLA Signed",
        "color": "e7e7e7",
        "default": false
      }
    ],
    "state": "open",
    "locked": false,
    "assignee": null,
    "assignees": [

    ],
    "milestone": null,
    "comments": 0,
    "created_at": "2019-02-28T11:01:15Z",
    "updated_at": "2019-02-28T11:09:49Z",
    "closed_at": null,
    "author_association": "COLLABORATOR",
    "pull_request": {
      "url": "https://api.github.com/repos/facebook/react/pulls/14974",
      "html_url": "https://github.com/facebook/react/pull/14974",
      "diff_url": "https://github.com/facebook/react/pull/14974.diff",
      "patch_url": "https://github.com/facebook/react/pull/14974.patch"
    },
    "body": "While playing around with the new `act()` API I discovered a test that\r\nwas unexpectedly failing. After some digging I noticed that the test is\r\npassing when run in isolation and that the test suite can be fixe by\r\nproperly calling `ReactDOM.unmountComponentAtNode()` to clean up after\r\nevery test run.\r\n\r\nYou can reproduce the issue by simply commenting out the call to\r\n`unmountComponentAtNode` in the `afterEach` callback. The newly added\r\ntest will fail but pass in isolation.\r\n\r\nI thought it would be a good idea to fix this upstream so nobody runs\r\ninto this issue again."
  },
  {
    "url": "https://api.github.com/repos/facebook/react/issues/15030",
    "repository_url": "https://api.github.com/repos/facebook/react",
    "labels_url": "https://api.github.com/repos/facebook/react/issues/15030/labels{/name}",
    "comments_url": "https://api.github.com/repos/facebook/react/issues/15030/comments",
    "events_url": "https://api.github.com/repos/facebook/react/issues/15030/events",
    "html_url": "https://github.com/facebook/react/pull/15030",
    "id": 417620384,
    "node_id": "MDExOlB1bGxSZXF1ZXN0MjU4NTc2NzQx",
    "number": 15030,
    "title": "Test suspended children are hidden before layout in persistent mode",
    "user": {
      "login": "acdlite",
      "id": 3624098,
      "node_id": "MDQ6VXNlcjM2MjQwOTg=",
      "avatar_url": "https://avatars0.githubusercontent.com/u/3624098?v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/acdlite",
      "html_url": "https://github.com/acdlite",
      "followers_url": "https://api.github.com/users/acdlite/followers",
      "following_url": "https://api.github.com/users/acdlite/following{/other_user}",
      "gists_url": "https://api.github.com/users/acdlite/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/acdlite/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/acdlite/subscriptions",
      "organizations_url": "https://api.github.com/users/acdlite/orgs",
      "repos_url": "https://api.github.com/users/acdlite/repos",
      "events_url": "https://api.github.com/users/acdlite/events{/privacy}",
      "received_events_url": "https://api.github.com/users/acdlite/received_events",
      "type": "User",
      "site_admin": false
    },
    "labels": [
      {
        "id": 196858374,
        "node_id": "MDU6TGFiZWwxOTY4NTgzNzQ=",
        "url": "https://api.github.com/repos/facebook/react/labels/CLA%20Signed",
        "name": "CLA Signed",
        "color": "e7e7e7",
        "default": false
      }
    ],
    "state": "open",
    "locked": false,
    "assignee": null,
    "assignees": [],
    "milestone": null,
    "comments": 1,
    "created_at": "2019-03-06T04:43:07Z",
    "updated_at": "2019-03-06T04:56:15Z",
    "closed_at": null,
    "author_association": "MEMBER",
    "pull_request": {
      "url": "https://api.github.com/repos/facebook/react/pulls/15030",
      "html_url": "https://github.com/facebook/react/pull/15030",
      "diff_url": "https://github.com/facebook/react/pull/15030.diff",
      "patch_url": "https://github.com/facebook/react/pull/15030.patch"
    },
    "body": "## Based on #15029 and #15013\r\n\r\nRefs behave differently in persistent mode, so instead of a ref, the persistent mode version of this test asserts on the output of the host tree."
  }
]