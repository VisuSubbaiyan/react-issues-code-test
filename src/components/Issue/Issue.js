import React from 'react';
// import PropTypes from 'prop-types';
import { IssueWrapper, IssueContainer } from './Issue.styled';
import Label from '../Label';
import Link from '../Link';
import Comments from '../Comments';
import { StatusIcon } from '../Icons';
import { deriveOlderDays } from '../../helpers/util';

const renderLabels = labels => labels.map(({ id, name, color }) =>
  <Label key={id} bgColor={`#${color}`} fontWeight='bold' chips >{name}</Label>);

const additionalCss = `font-size: 14px; font-weight: bold; line-height: 14px; vertical-align: text-top;`;

const Issue = ({ title, html_url, labels = [], comments, number, created_at, user }) => (
  <IssueWrapper className="issue">
    <StatusIcon fill="green" css={`flex: 1 0 0;`} />
    <IssueContainer>
      <Link href={html_url} css={additionalCss}>{title}</Link>
      {renderLabels(labels)}
      <IssueContainer>
        <Label>{`#${number} opened ${deriveOlderDays(created_at)} ago by ${user.login}`}</Label>
      </IssueContainer>
    </IssueContainer>
    <Comments comments={comments} href={html_url} />
  </IssueWrapper>
);

export default Issue;
