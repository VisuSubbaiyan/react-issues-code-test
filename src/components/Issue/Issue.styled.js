import styled from 'styled-components';
import { viewport, styling } from '../../helpers/util';

export const IssueWrapper = styled.div`
  margin: 10px 0;
  border: 1px solid ${props => props.theme.secondary};
  border-radius: 3px;
  padding: 5px;
  background: #fff;
  min-height: 60px;
  display: flex;
  justify-content: space-between;
  
  ${styling}
`;

export const IssueContainer = styled.div`
  flex: 12 0 0;
  ${props => viewport(props.theme.breakpoints.small)`flex: 18 0 0;`}
`;
