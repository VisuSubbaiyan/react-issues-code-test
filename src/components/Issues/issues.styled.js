import styled from 'styled-components';
import { viewport } from '../../helpers/util';

export const IssuesMainWrapper = styled.div`
  padding-top: 60px;
  width: 90%;
  margin: 0 auto;
  ${props => viewport(props.theme.breakpoints.medium)`width: 720px;`}
`;

export const SpinnerWrapper = styled.div`
  display: flex;
  min-height: 40px;
  justify-content: center;
`;
