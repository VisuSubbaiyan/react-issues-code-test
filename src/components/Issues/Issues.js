import React, { Component } from 'react';
import githubAPI from '../../services/githubServices';
import Issue from '../Issue';
import { IssuesMainWrapper, SpinnerWrapper } from './issues.styled';
import Spinner from '../Spinner';

const renderIssues = (issues = []) => issues.map(issue => <Issue {...issue} key={issue.id} />);

class Issues extends Component {
  static defaultProps = {
    issues: []
  }

  state = {
    page: 1,
    status: 'open',
    loading: false,
    issues: this.props.issues,
    allIssuesLoaded: false
  }

  componentDidMount() {
    this.loadIssues();
    window.addEventListener('scroll', e => this.handleScrolling(e));
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', e => this.handleScrolling(e));
  }

  handleScrolling = (e) => {
    const { loading, allIssuesLoaded } = this.state;
    if (loading || allIssuesLoaded) return;

    const lastChild = document.querySelector('div.issues > div.issue:last-child') || {};
    const lastIssueOffset = (lastChild.offsetTop + lastChild.clientHeight) - 60;
    const windowOffset = window.pageYOffset + window.innerHeight;
    if (windowOffset > lastIssueOffset) {
      this.loadMoreIssues();
    }
  }

  loadIssues = async () => {
    const { page, status, issues } = this.state;
    const newIssues = await githubAPI
      .get('/issues', { params: { page, status } })
      .then(res => res.data);
    const allIssuesLoaded = newIssues.length === 0;

    this.setState({ loading: false, issues: [...issues, ...newIssues], allIssuesLoaded });
  }

  loadMoreIssues = () => {
    this.setState(prevState => ({
      page: prevState.page + 1,
      loading: true,
    }), this.loadIssues);
  }

  render() {
    const { loading, issues } = this.state;

    return (
      <IssuesMainWrapper className="issues">
        {renderIssues(issues)}
        {loading && <SpinnerWrapper><Spinner /></SpinnerWrapper>}
      </IssuesMainWrapper>
    )
  }
}

export default Issues;
