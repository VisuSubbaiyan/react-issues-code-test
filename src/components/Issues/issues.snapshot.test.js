import React from 'react';
import 'jest-styled-components';

import Issues from './Issues';
import Issue from '../Issue';
import { createMountedComponent, sampleIssues } from '../../helpers/testUtil';

describe('Issues should', () => {
  it('render correctly', () => {
    const mountedIssues = createMountedComponent({
      children: <Issues issues={sampleIssues} />,
    });

    expect(mountedIssues.htmlRender()).toMatchSnapshot();
    expect(mountedIssues.find(Issue)).toHaveLength(2);
  });
});