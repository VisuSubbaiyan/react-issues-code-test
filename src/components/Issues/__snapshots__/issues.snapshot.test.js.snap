// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issues should render correctly 1`] = `
"<div class=\\"issues sc-bZQynM kueXeN\\">
  <div class=\\"issue sc-bdVaJa IwQOz\\">
    <svg fill=\\"green\\" css=\\"flex: 1 0 0;\\" class=\\"sc-EHOje isaVFh\\">issueStatus.svg</svg>
    <div class=\\"sc-bwzfXH jXZiCh\\"><a href=\\"https://github.com/facebook/react/pull/14974\\" class=\\"sc-bxivhb KqdGy\\">Fix act() tests by properly unmounting the container</a>
      <label
      font-weight=\\"bold\\" class=\\"sc-htpNat kPKryi\\">CLA Signed</label>
        <div class=\\"sc-bwzfXH jXZiCh\\">
          <label font-weight=\\"normal\\" class=\\"sc-htpNat ePqaPb\\">#14974 opened 6 days ago by philipp-spiess</label>
        </div>
    </div>
  </div>
  <div class=\\"issue sc-bdVaJa IwQOz\\">
    <svg fill=\\"green\\" css=\\"flex: 1 0 0;\\" class=\\"sc-EHOje isaVFh\\">issueStatus.svg</svg>
    <div class=\\"sc-bwzfXH jXZiCh\\"><a href=\\"https://github.com/facebook/react/pull/15030\\" class=\\"sc-bxivhb KqdGy\\">Test suspended children are hidden before layout in persistent mode</a>
      <label
      font-weight=\\"bold\\" class=\\"sc-htpNat kPKryi\\">CLA Signed</label>
        <div class=\\"sc-bwzfXH jXZiCh\\">
          <label font-weight=\\"normal\\" class=\\"sc-htpNat ePqaPb\\">#15030 opened about 18 hours ago by acdlite</label>
        </div>
    </div><a href=\\"https://github.com/facebook/react/pull/15030\\" class=\\"sc-bxivhb kosjtV\\"><svg class=\\"sc-ifAKCX cFlEyZ\\">bubble.svg</svg><span>1</span></a>
  </div>
</div>"
`;
